exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatchSizer.java

```diff
@@ -76,110 +82,327 @@
+    public boolean isVariableWidth() {
+      return isVariableWidth;
+    }
 
     public Map<String, ColumnSize> getChildren() {
       return children;
     }
 
+    public boolean isComplex() {
+      if (metadata.getType().getMinorType() == MinorType.MAP ||
+        metadata.getType().getMinorType() == MinorType.UNION ||
+        metadata.getType().getMinorType() == MinorType.LIST) {
+        return true;
+      }
+      return false;
+    }
+
+    public boolean isRepeatedList() {
+      if (metadata.getType().getMinorType() == MinorType.LIST &&
+        metadata.getDataMode() == DataMode.REPEATED) {
+        return true;
+      }
+      return false;
+    }
+
+    /**
+     * This is the average per entry width, used for vector allocation.
+     */
+    public int getEntryWidth() {
+      int width = 0;
+      if (isVariableWidth) {
+        width = getNetSizePerEntry() - OFFSET_VECTOR_WIDTH;
+
+        // Subtract out the bits (is-set) vector width
+        if (metadata.getDataMode() == DataMode.OPTIONAL) {
+          width -= BIT_VECTOR_WIDTH;
+        }
+      }
+
+      return (safeDivide(width, estElementCountPerArray));
+    }
+
     public ColumnSize(ValueVector v, String prefix) {
       this.prefix = prefix;
       valueCount = v.getAccessor().getValueCount();
       metadata = v.getField();
-      isVariableWidth = v instanceof VariableWidthVector;
-
-      // The amount of memory consumed by the payload: the actual
-      // data stored in the vectors.
-
-      if (v.getField().getDataMode() == DataMode.REPEATED) {
-        elementCount = buildRepeated(v);
-        estElementCountPerArray = valueCount == 0 ? 0 : elementCount * 1.0f / valueCount;
-      } else {
-        elementCount = 1;
-        estElementCountPerArray = 1;
+      isVariableWidth = (v instanceof VariableWidthVector || v instanceof RepeatedVariableWidthVectorLike);
+      elementCount = valueCount;
+      estElementCountPerArray = 1;
+      totalNetSize = v.getPayloadByteCount(valueCount);
+
+      // Special case. For union and list vectors, it is very complex
+      // to figure out raw data size. Make it same as net size.
+      if (metadata.getType().getMinorType() == MinorType.UNION ||
+        (metadata.getType().getMinorType() == MinorType.LIST && v.getField().getDataMode() != DataMode.REPEATED)) {
+        totalDataSize = totalNetSize;
       }
-      switch (metadata.getType().getMinorType()) {
-      case LIST:
-        buildList(v);
-        break;
-      case MAP:
-      case UNION:
-        // No standard size for Union type
-        dataSize = v.getPayloadByteCount(valueCount);
-        break;
-      default:
-        dataSize = v.getPayloadByteCount(valueCount);
-        try {
-          stdSize = TypeHelper.getSize(metadata.getType()) * elementCount;
-        } catch (Exception e) {
-          // For unsupported types, just set stdSize to 0.
-          stdSize = 0;
-        }
+
+      switch(v.getField().getDataMode()) {
```

Comment:
Once code gets this complex, it may be time to create subclasses for the three modes. Plus subclasses for the categories of types (variable-width, map, etc.)

The column size would be a composition of the type size class plus a cardinality size class.

Maybe overkill for the current round of changes, but something to keep in mind if this code continues to grow in complexity.